core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -988,4 +990,15 @@ public static SqlCall stripSeparator(SqlCall call) {
       return relDataType;
     }
   };
+
+  public static final SqlReturnTypeInference PERCENTILE_DISC = opBinding -> {
+    if (opBinding instanceof SqlWithinGroupOperator.PercentileDiscCallBinding) {
```

Comment:
This 'if .. instanceof ... else if ... instanceof' pattern isn't good. How about adding a method `getCollationExpression()` to `SqlCallBinding`, with a default implementation that throws. Then you can remove both `getCollationType` methods.

Reply:
I can't implement `getCollationExpression` in `PercentileDiscAggCallBinding`, as I don't have the collation expression there. 
(maybe i'm missing something). I can however implement `getCollationType` which will remove the need to use the `instance of` pattern here.